### ISSUE TITLE
Distributed Semantic MediaWiki

### ISSUE THREAD

**Issue by mwjames** [SemanticMediaWiki/SemanticMediaWiki#796]
### Definition 
The understanding is that a "Distributed Semantic MediaWiki" is to reroute storage and query computation of property/values to a separated system which is physical or logical different from the actual wiki interface a user gets in contact.

- Do we want to support such infrastructural capability of  Semantic MediaWiki?
- Why would we want such feature?
- Can Semantic MediaWiki be made to be used as a distributed service?
- What would be the benefit to support a "Distributed Semantic MediaWiki"?
- What core changes are necessary and how would such service need to be setup?
- What scenarios can be supported by a "Distributed Semantic MediaWiki" and which scenarios cannot be supported?
- What are possible pitfalls and immobilizing scenarios that would make "Distributed Semantic MediaWiki" a rather incomprehensible complex endeavour?

### Some preliminary comments
- Add store router class to redirect to a store-endpoint
- Stronger separation of Client/Server components within the SMW namespace

Keywords: Master/Slave, Loadbalancer, Client/Server

PS: Semantic MediaWiki already supports `SPARQLStore` as an external service where query computation is distributed to a different source that is not within the same wiki vicinity (DB server access) .

refs #757

**Comment by mwjames** [comment]
@kghbln @thingles @JeroenDeDauw @mkroetzsch @s7eph4n I'd be grateful for some enlightening comments to above questions.

**Comment by s7eph4n** [comment]
Load balancing and master/slave constructs have never been relevant for me, so I can not say much about it.

Is there actually demand for such an ability? I know, that demand often arises from the availability of a feature, but it might be useful to get in contact with the owners of large installations. I remember SNPedia. WikiApiary should have a considerable size, as well. There are others, IIRC.

Another use case might be for several wikis to share a common backend, although again I am not sure who might actually use that.

A first approach to an implementation could probably be to just simply MW+SMW on both servers and use the API. In the next step you could then have some SMW-only server. No idea, how hard it would actually be to strip down the code. On SMW-side probably not too hard, on MW-side it could be easier to just reimplement the API code.

By using the already available API you would also enable some not-yet-available JS-based result formats (or any other software) to access the SMW endpoint. In the end you would have a server similar in functionality to SPARQL, only with a less powerful, but simpler query language.

**Comment by JeroenDeDauw** [comment]
I'm not sure I understand what the intend would be here. There is talk about load balancing and performance. There used to be this "Distributed Semantic MediaWiki" project, which was about having the data on multiple places, in the same direction as the [Push extension](http://www.mediawiki.org/wiki/Extension:Push), and not about performance at all. Those two goals are very different, so understanding about which one this is about is important.

If we want distribution for non performance reasons, I suggest to look at [the federated wiki stuff](https://github.com/fedwiki). It seems like some very interesting things are happening there, and that some of the innovations they already made could be brought to SMW. I've been wanting to have a real good look at it for some time now but have not managed so far.

**Comment by mwjames** [comment]
> I'm not sure I understand what the intend would be here. There is talk about load balancing and performance.

It is about "to reroute storage and query computation of property/values" which means instead of having SMW doing all the work on the one server that provides the wiki engine to provide a "distributive" feature that would allow to entangle those and have the non-compute intensive parts on the wiki engine while the query/storage can be placed on a dedicated server that does not require a public (from a user perspective ) access or is only accessible via an API. (similar to the current `SPARQLStore` approach but for the works of the `SQLStore`). In short it is to distribute possible performance bottlenecks caused by query execution on behalf of the `SQLStore\QueryEngine/StorageEngine`.

The main question is whether such a feature would be necessary for SMW to support or not. If we come to the conclusion it is not feasible or it is too much work to start to begin with then this is fine as well. 

> If we want distribution for non performance reasons, I suggest to look at the federated wiki stuff. It seems like some very interesting things are happening there

"federated wiki stuff" would require a non-MediaWiki API? (I'm not saying this would be a bad thing) 

PS: Whether it is called "Distributed Semantic MediaWiki" or something else is of a secondary nature (I just needed a working title).

**Comment by JeroenDeDauw** [comment]
Ok, so it's about performance. Then I'm with @s7eph4n wondering if there is need for this.

You seem to be starting with a solution looking for performance problems to tackle, rather than with observed performance problems, and then determining what the most applicable solutions are. IMO the later approach makes  a lot more sense.

> "federated wiki stuff" would require a non-MediaWiki API? (I'm not saying this would be a bad thing) 

We could implement some of their innovations on top of MediaWiki and current SMW code.

> PS: Whether it is called "Distributed Semantic MediaWiki" or something else is of a secondary nature (I just needed a working title).

This title is confusing for those that know [Distributed Semantic MediaWiki](http://www.mediawiki.org/wiki/Extension:DSMW)

**Comment by mwjames** [comment]
> Ok, so it's about performance. Then I'm with @s7eph4n wondering if there is need for this.

I'm wondering this myself but if we wait for the demand to pile up then we are already to late for the party.

> You seem to be starting with a solution looking for performance problems to tackle, rather than with observed performance problems, and then determining what the most applicable solutions are. IMO the later approach makes a lot more sense.

I'm locking for features that can be developed or at least can be taken into account when re-factoring efforts are made. In order to judge whether such scenarios are feasible or not one at least  needs to entertain those possibilities.

> A first approach to an implementation could probably be to just simply MW+SMW on both servers and use the API.

Certainly that would be the first step but since we can not reroute a request from server A to server B via an dedicated API we would already fail here.

The question would be of what needs to be done on the `SQLStore` side to actually allow for this kind of scenario. (we have the querysource parameter but that only works for the #ask query and only for the query and not for the storage).

Do we know whether a master-slave scenario does work with SMW for a pure DB-based solution (only to distribute data from A to B)? Internally (using the MW-DB layer) we read from slave and write to master but this isn't sufficient for distributing storage/query requests to another back-end. 

Wikia was experimenting [0] with something they called SMW cluster vs local DB.

[0] https://github.com/Wikia/app/pull/4917#issuecomment-57535495

**Comment by mkroetzsch** [comment]
I don't think that any SMW use case, current or future, is in the scale that would justify distribution (scale out). In the cases where I have been looking into SMWs with real performance issues (Wikia), the reason was database thrashing with small queries and PHP overloading with thousands of templates/parser functions per page. Slow SQL queries I have only seen in cases where the MySQL query optimizer has failed, and in these cases even tiny databases would produce the problem (a few thousand triples, yet the query would not come back even after minutes).

English Wikipedia and Wikidata each run on a single master DB server. The data we have is nowhere near the scalability limit of a disk-backed database management system. Just for context, I am using the Yaron's Miga viewer with around 50K records, each with ten or more property values as an in-browser app. Likewise, the Wikidata people viewer Gerog Wild created works with about 300K items, each with several property values in Javascript. Each of these uses less than 1G of memory. In comparison, smw.o has merely 90K property values in total.

Therefore, what I would do to improve performance would be to create an in-memory storage backend. It would run as a separate service (like a lightweight "database" but without the complete DBMS feature set) that is used to answer all SMW queries, lookups and ask alike. The service would be kept in synch with the wiki (be notified of all changes) but stored data would also be kept in SQL as before. If the in-memory service crashes, it would be easy to restart it from SQL or SQL could be used as a fallback. The overall design principle of the whole thing would be that all data can be kept in memory at all times. All smaller wikis could use this without concerns. Very large wikis would need to consider the memory requirements and use a sufficiently high-powered server (maybe give it 50G or memory instead of 5G). One could implement tree-shaped query computation to ensure that intermediate results during query answering are in the order of the overall wiki size, so that basically any query could be executed in memory given enough time. The key thing is to avoid PHP since it is incredibly memory inefficient.

Nevertheless, such an in-memory system would mainly improve "slow query" performance and enable advanced interactions (e.g., real-time faceted browsing). It would not help much with thrashing effects and slow PHP parsing. Thrashing would be addressed by bundling storage requests, to avoid thousands of calls to the API of the in-memory system.

One consideration before starting such a project would be whether one could use an existing in-memory storage for handling queries, and merely write a wrapper service to provide a fast, SMW-tailored API that can load data from SQL and implement special functions such as bulk-fetching of small query results needed for ask result display. Downside would be that configuration would be more work if users would have to install something like Virtuoso or another additional DBMS first.

Smaller improvements of performance on the current architecture would probably start with the better use of memcached in our storage layers. Currently, we keep things like page id cashes in PHP data structures, which is incredibly inefficient (but reduces DB thrashing significantly).

**Comment by JeroenDeDauw** [comment]
Thanks for the write up @mkroetzsch. It's much along the lines of what I was thinking. Only somewhat related: do you know Wikia is still running SMW 1.7.1? :)

@mwjames the relevant buzzword for what you are describing seems to be "Microservices". At least, that is what I start thinking about.

Having a separate storage and a separate query service might make sense in some situations. And once you have dedicated components for these tasks, it's not hard to enable those to operate in different processes or on different servers. Take for instance Wikibase QueryEngine, which is essentially doing the same thing as the SQLStore. Since it does not depend on tables or configuration not part of the component, and nothing outside of the component depends on its tables, you can easily have it run somewhere else. Or run a pile of instances on different servers and load balance between those behind some simple facade.

Having good boundaries/components is a prerequisite for "microservices". On the road to getting  good boundaries/components, we'll be able to nicely add in other performance related improvements, such as adding in effective caching layers, or avoiding execution of code that is not required.

**Comment by mwjames** [comment]
I always appreciate the feedback, now we have at least something to point to in case someone asks or can be used as input for further discussions.
